### PR TITLE
Add support for gathering profiles and delivering them to Pyroscope

### PR DESCRIPTION
**Pull request by petewall**
If `profiles.enabled=true`, then we'll deploy a Grafana Agent instance as a daemonset and it scrapes profiles from annotated pods as well as using ebpf.

Test code is updated to deploy OSS Pyroscope to the test cluster, and a query was added to one of the values file to check that profiles exists in the db.

Reviewer notes:
* No examples were touched except for the new `examples/profiles-enabled`. The only difference between that examlpe and default is the enabling of the profiles feature. You can ignore all example changes except maybe the output.yaml and profiles.river for the new example.
* The [values.yaml](https://github.com/grafana/k8s-monitoring-helm/pull/446/files#diff-cd22d083c6f49784d198f55b990feb9ae245abf1353dd56a379ec880126f4488) has the most meaningful changes, since that's the "user interface" for this feature. Make sure it'd make sense for your users.
* Feel free to ignore `.github/configs/updatecli.d` It's the Helm chart's dependabot
* The generation of the agent config for profiles is split into 3 files:
  * Gathering and processing pprof profiles from annotations: https://github.com/grafana/k8s-monitoring-helm/pull/446/files#diff-a7d5d9e9827dc0a411c8c2a1853feb9eb2bdb5abf982e767b6505b84970dc0db (this was stolen from the agent config from the pyroscope helm chart)
  * Gathering and processing eBPF profiles: https://github.com/grafana/k8s-monitoring-helm/pull/446/files#diff-a864bb4baeae9e3b5795da963df14e43811503f6ecca62016db891121d2dd1b7 
  * Delivering profiles to Pyroscope: https://github.com/grafana/k8s-monitoring-helm/pull/446/files#diff-26b4fa3b600a3276456cf3753879c42270cc6e3ea97554b604c634ce5e819f48
*

### CODE REVIEW

File: scripts/setup-local-test-cluster.sh
Comment (simonswine) [inline scripts/setup-local-test-cluster.sh:67]: By default this will not support multiple tenants. You will need to supply a extraArg like that:

```
--set pyroscope.extraArgs."auth.multitenancy-enabled"=true
```

**Review by simonswine — COMMENTED** [review]
Generally looking good, will need to give a deeper look later.

File: .github/configs/grafana.yaml
Comment (petewall) [inline .github/configs/grafana.yaml:37]: Pre-configuring the test pyroscope data source in our test grafana instance.

File: .github/configs/pyroscope.yaml
Comment (petewall) [inline .github/configs/pyroscope.yaml:5]: This is how we're deploying the test pyroscope

File: charts/k8s-monitoring/test/query-test.sh
Comment (petewall) [inline charts/k8s-monitoring/test/query-test.sh:183]: Does Pyroscope have a name for its query language? `ProfileQL`? `PyroQL`? The literal 🔥 emoji?

Comment (simonswine) [reply charts/k8s-monitoring/test/query-test.sh:183]: So far we don't have a language other than the label selectors that we handle like LogQL/PromQL, but I think `profileql` or `pyroql` is good enough for those purposes. We can obvisouly also go on a UTF-8 journey with the emoji 😆

File: charts/k8s-monitoring/values.yaml
Comment (simonswine) [inline charts/k8s-monitoring/values.yaml:855]:
```suggestion
    # -- Gather profiles by scraping pprof HTTP endpoints
```

File: charts/k8s-monitoring/templates/agent_config/_profiles_ebpf.river.txt
Comment (simonswine) [inline charts/k8s-monitoring/templates/agent_config/_profiles_ebpf.river.txt:48]: I do think it makes more sense without the `k8s/`  "folder".

```suggestion
    replacement = "ebpf/${1}/${2}"
```

File: charts/k8s-monitoring/values.yaml
Comment (simonswine) [inline charts/k8s-monitoring/values.yaml:867]: I do think `extraRelabelingRules` also make sense here

File: charts/k8s-monitoring/templates/agent_config/_profiles_service.river.txt
Comment (simonswine) [inline charts/k8s-monitoring/templates/agent_config/_profiles_service.river.txt:17]: That seems like a copy paste problem 🙂:

```suggestion
    username = nonsensitive(remote.kubernetes.secret.profiles_service.data[{{ .basicAuth.usernameKey | quote }}])
    password = remote.kubernetes.secret.profiles_service.data[{{ .basicAuth.passwordKey | quote }}]
```

**Review by simonswine — APPROVED** [review]
Thank you for your work on this! I think this is almost there.

I found a few smaller issues, that need fixing, once they are fixed no concerns with merging. Almost LGTM 😆

File: charts/k8s-monitoring/templates/agent_config/_profiles_service.river.txt
Comment (petewall) [reply charts/k8s-monitoring/templates/agent_config/_profiles_service.river.txt:17]: Yes! Thanks for the catch!

File: charts/k8s-monitoring/templates/agent_config/_profiles_ebpf.river.txt
Comment (simonswine) [inline charts/k8s-monitoring/templates/agent_config/_profiles_ebpf.river.txt:61]: I think for this it would be good if we could expose `demangle` as a value, with the default value "none". In some cases it needs to be set to `"full"`.

```suggestion
  forward_to = [pyroscope.write.profiles_service.receiver]
  demangle = {{ .Values.profiles.ebpf.demangle }}"
}
```

File: charts/k8s-monitoring/templates/agent_config/_profiles_service.river.txt
Comment (simonswine) [inline charts/k8s-monitoring/templates/agent_config/_profiles_service.river.txt:26]: Basic auth needs to be under endpoint:

```suggestion
{{ if eq .authMode "basic" }}
    basic_auth {
      username = nonsensitive(remote.kubernetes.secret.profiles_service.data[{{ .basicAuth.usernameKey | quote }}])
      password = remote.kubernetes.secret.profiles_service.data[{{ .basicAuth.passwordKey | quote }}]
    }
{{- end }}
{{- if .tls }}
    tls_config {
    {{- range $k, $v := .tls }}
      {{ $k }} = {{ $v | toJson }}
    {{- end }}
    }
{{- end }}
  }
```

File: charts/k8s-monitoring/values.yaml
Comment (simonswine) [inline charts/k8s-monitoring/values.yaml:1318]: This is important for eBPF to work

```suggestion
    type: daemonset
    hostPID: true
```

**Review by simonswine — COMMENTED** [review]
Thanks for addressing so many things already, unfortunately a second review found some more 😆.

I think that's hopefully everything

File: charts/k8s-monitoring/templates/agent_config/_profiles_pprof.river.txt
Comment (skl) [inline charts/k8s-monitoring/templates/agent_config/_profiles_pprof.river.txt:55]: Just for my own understanding, what does this syntax mean: `{{.}}`

File: charts/k8s-monitoring/templates/agent_config/_profiles_ebpf.river.txt
Comment (skl) [inline charts/k8s-monitoring/templates/agent_config/_profiles_ebpf.river.txt:21]: Do you need `Pending` here as well? For consistency with `discovery.relabel.pprof_pods`

**Review by skl — APPROVED** [review]
🚀

File: charts/k8s-monitoring/templates/agent_config/_profiles_ebpf.river.txt
Comment (petewall) [reply charts/k8s-monitoring/templates/agent_config/_profiles_ebpf.river.txt:21]: @simonswine ?

File: charts/k8s-monitoring/templates/agent_config/_profiles_pprof.river.txt
Comment (petewall) [reply charts/k8s-monitoring/templates/agent_config/_profiles_pprof.river.txt:55]: Typically. `.` is the top level data object. so `.Values` means delve inside the values `.Release` goes into Helm's pre-defined release object.

The `range` command iterates over a list and resets `.` to each item in that list. That's also why sometimes you see saving a variable before the range, as in with
```
$profileTypes := .Values.profiles.pprof.types
```
because that way we can still reference that item even though `.` has become something else.


You'll also see things like
```
{{ with .Values.externalServices.prometheus }}
```
which resets `.` to be starting at `.Values.externalServices.prometheus`.

Comment (petewall) [reply charts/k8s-monitoring/templates/agent_config/_profiles_pprof.river.txt:55]: In this case, `{{ . }}` is referencing the current element in the `.Values.profiles.pprof.types` array

File: charts/k8s-monitoring/templates/agent_config/_profiles_ebpf.river.txt
Comment (simonswine) [reply charts/k8s-monitoring/templates/agent_config/_profiles_ebpf.river.txt:21]: No, having `Pending` ones included is fine for ebpf.

Although I noticed that according to https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase Completed is not a valid state

**Review by simonswine — APPROVED** [review]
Tested this e2e again! All looking good to me. Thanks for getting this together.

LGTM

**Pull request merged**
